src/libstat/backends/redis_backend.c

```diff
@@ -912,30 +920,50 @@ rspamd_redis_stat_keys (redisAsyncContext *c, gpointer r, gpointer priv)
 			}
 		}
 
-		/* Set up the required keys */
-		ucl_object_insert_key (cbdata->cur,
-				ucl_object_typed_new (UCL_INT), "revision", 0, false);
-		ucl_object_insert_key (cbdata->cur,
-				ucl_object_typed_new (UCL_INT), "used", 0, false);
-		ucl_object_insert_key (cbdata->cur,
-				ucl_object_typed_new (UCL_INT), "total", 0, false);
-		ucl_object_insert_key (cbdata->cur,
-				ucl_object_typed_new (UCL_INT), "size", 0, false);
-		ucl_object_insert_key (cbdata->cur,
-				ucl_object_fromstring (cbdata->elt->ctx->stcf->symbol),
-				"symbol", 0, false);
-		ucl_object_insert_key (cbdata->cur, ucl_object_fromstring ("redis"),
-				"type", 0, false);
-		ucl_object_insert_key (cbdata->cur, ucl_object_fromint (0),
-				"languages", 0, false);
-		ucl_object_insert_key (cbdata->cur, ucl_object_fromint (processed),
-				"users", 0, false);
-
-		rspamd_upstream_ok (cbdata->selected);
-
-		if (cbdata->inflight == 0) {
-			rspamd_redis_async_cbdata_cleanup (cbdata);
-			redis_elt->cbdata = NULL;
+		if (more) {
+			/* Cleanup the cbdata->cur_keys and re-allowcate */
+			for (i = 0; i < cbdata->cur_keys->len; i ++) {
+				k = g_ptr_array_index (cbdata->cur_keys, i);
+				g_free (k);
+			}
+
+			g_ptr_array_free (cbdata->cur_keys, TRUE);
```

Comment:
Why do you free these objects? Reallocating memory seems completely useless for me - you can just set its size to zero. I'm even not quite sure that we need to dup keys in `cur_keys` as redisAsyncCommand already copies all stuff internally. Array reallocation is definitely redundant.

Reply:
Agree, I myself also wonder why we have to store the key into `cur_keys`.

Then just remove the cleanup code, and let the `g_ptr_array_set_size ` in here to do the job:

https://github.com/rspamd/rspamd/blob/fbd2c0e1e3da12793dbf2681feac1244e9ae5485/src/libstat/backends/redis_backend.c#L863